node/src/components/consensus/highway_core/state/endorsement_tracker.rs

```diff
@@ -0,0 +1,103 @@
+use std::collections::HashMap;
+
+use crate::components::consensus::{highway_core::validators::ValidatorIndex, traits::Context};
+
+use super::{Panorama, State};
+
+/// Helper structure used for computing which votes need endorsement.
+pub(crate) struct EndorsementTracker<'a, C: Context> {
+    /// Validators that are known to be equivocators by the initial panorama.
+    known_equivocators: Vec<ValidatorIndex>,
+    /// Visited units. Used for avoding visitng the same unit twice.
+    visited_nodes: Vec<C::Hash>,
```

Comment:
Why not make this a set? We're almost exclusively using this for membership tests.

Reply:
I thought it will be better performance since set will do the hash and compare.